src/google/credential.rs

```diff
@@ -1,24 +1,34 @@
+pub mod external_account;
+pub mod service_account;
+
 use std::env;
 use std::sync::Arc;
 use std::sync::Mutex;
 
+use anyhow::bail;
 use anyhow::Result;
 use log::debug;
-use serde::Deserialize;
+
+pub use self::external_account::ExternalAccount;
+pub use self::service_account::ServiceAccount;
 
 use super::constants::GOOGLE_APPLICATION_CREDENTIALS;
 use crate::hash::base64_decode;
 
-/// Credential is the file which stores service account's client_id and private key.
-#[derive(Clone, Deserialize)]
+/// A Google credential account, either a service or an external account.
+#[derive(Clone, serde::Deserialize)]
 #[cfg_attr(test, derive(Debug))]
-pub struct Credential {
-    /// Private key of credential
-    pub private_key: String,
-    /// The client email of credential
-    pub client_email: String,
+#[serde(rename_all = "snake_case", tag = "type")]
+pub enum CredentialAccount {
```

Comment:
I think `Credential` suits it better if you are willing to do a breaking change for the next release

Reply:
I prefer to keep `Credential` and add a new `ty` (renamed as `type`) field instead.